### ISSUE TITLE
game id support

### ISSUE THREAD

**Issue by ZjemCiKolege** [DKWDRV/DKWDRV#58]
It would be nice if dkwdrv  send info for sd2psx 🥉 

https://github.com/ps2-mmce/mmceman

**Comment by DKWDRV** [comment]
Added in TODO list with low priority due to lack of device.

**Comment by DKWDRV** [comment]
Making this higher priority. Do we have a fast way to get this working? How will the id be send to the game? Before the boot? PGIF and DECKARD models too?
In case of PSX.EXE games how will the checksum be sent?

**Comment by pcm720** [comment]
@DKWDRV 
> Do we have a fast way to get this working? 

There are two ways to trigger VMC switch on SD2PSX/MCP2:
1. Writing to the history file the same way the original OSD does
2. Loading mmceman.irx and using devctl call to mount the VMC

The first way is implemented here:
https://github.com/pcm720/osdmenu-launcher/blob/main/launcher/src/history.c

The second way is implemented here:
https://github.com/pcm720/nhddl/blob/8863e1ba5f5691211a77b1837afa15365171d596/src/devices_mmce.c#L43-L75

You can use my code if you want/need to, I don't mind.

> In case of PSX.EXE games how will the checksum be sent?

Both ways only require the title ID, so you don't need any checksums.
As far as I'm aware, DKWDRV already detects the title ID during the disc/ISO scan, so you're all set for games that don't have the SYSTEM.CNF or use the default executable name.

**Comment by DKWDRV** [comment]
What way do users prefer? Maybe they want history too? For both CD and USB?
What if they want to load VMC but no history?

DKWDRV only checks for some titles, the whole list would be too much. For the rest is calculates the CRC of PSX.EXE and uses that as ID.

**Comment by pcm720** [comment]
> What way do users prefer? Maybe they want history too? For both CD and USB?
What if they want to load VMC but no history?

It doesn't make sense to care about history on MMCE devices because the VMC is constantly switching...
And considering that this history isn't used anywhere other than the OSDSYS intro, I personally went with the mmceman route.

The only reason osdmenu-launcher uses the history file is to replicate the original behavior since it's used to load the disc from the hacked OSDSYS.

> DKWDRV only checks for some titles, the whole list would be too much. For the rest is calculates the CRC of PSX.EXE and uses that as ID.

You might find this useful then as MMCE devices require the title ID...
https://github.com/pcm720/osdmenu-launcher/blob/9e33cd287e5647ed8145227c89f5c758c27ffebc/launcher/src/handler_cdrom.c#L290

**Comment by DKWDRV** [comment]
There are more games with PSX.EXE than those. Maybe it can support ID starting with "PSX" from the [configs](https://github.com/DKWDRV/DKWDRV?tab=readme-ov-file#user-configs)

**Comment by pcm720** [comment]
> There are more games with PSX.EXE than those.

Commercial titles?
Would be nice to have a list of those then.

**Comment by DKWDRV** [comment]
> > There are more games with PSX.EXE than those.
> 
> Commercial titles?

Yes. At least 150+ titles. Atm do not have full list. Why not PSX checksum then?

**Comment by pcm720** [comment]
> Yes. At least 150+ titles. Atm do not have full list. Why not PSX checksum then?

Because MMCE devices need the title ID and the RetroGEM uses it for per-game profiles...

SD2PSX in particular uses this database to show the name instead of the ID:
https://github.com/niemasd/GameDB-PSX

I've just noticed that it can be used for matching volume creation date with the title ID since it has both executable names and creation date (listed as UUID). I'll parse through it and see how many commercial titles have PSX.EXE.

**Comment by pcm720** [comment]
Just ran some quickly thrown together program through the PSX.data.json:

```
Total count of titles with PSX.EXE is 502
Total count of titles with valid title ID is 432
Total count of valid demo discs: 154, invalid: 3
Total count of valid sampler discs: 12, invalid: 0
Total count of valid unlicensed discs: 1, invalid: 41

Total count of duplicate creation dates (among PSX.EXE titles): 0
```

Some of entries without a valid title ID, like Toushinden and A Ressha de Ikou 4 Rev 1, are present in my table.
Some of valid entries are re-releases with different title ID (e.g. Arc the Lad).

[PS1_stats.txt](https://github.com/user-attachments/files/19352995/PS1_stats.txt)

**Comment by awaken1ng** [comment]
> There are two ways to trigger VMC switch on SD2PSX/MCP2:

Another one, for PS1 games in particular, is [MCP1 protocol](https://gitlab.com/chriz2600/ps1-game-id-transmission), sd2psXtd implementation can be seen [here](https://github.com/sd2psXtd/firmware/blob/38cf77377bff82a845e54ad91e76c7cd16a4fa6e/src/ps1/ps1_memory_card.c#L425).

**Comment by DKWDRV** [comment]
Does mmceman support all types of mc emus out there?
How big is? Stripped down version available?
Do we have author permission to integrate it in this project?
Does it interfere with normal SIO2MAN, MCMAN drivers? Any special precautions for it?

I have no device to test the implementation so we have to depend on already known code.

@pcm720 using your PS1_stats.txt for getting real ids nothing is compromised? [DECKARD config](https://github.com/DKWDRV/DKWDRV/blob/main/docs/files/config_deckard.md)

The timestamp strings will increase the filesize a lot unless they can be further shortened down.

**Comment by pcm720** [comment]
> Does mmceman support all types of mc emus out there?
How big is? Stripped down version available?
Do we have author permission to integrate it in this project?

All memory card emulation devices implement MMCE protocol.
https://github.com/ps2-mmce/mmceman

> Does it interfere with normal SIO2MAN, MCMAN drivers? Any special precautions for it?

AFAIK, it might be incompatible with the earlier SIO2MAN revisions from protokernels(?).
You'll be better off asking these questions in PS2 Scene Discord since all MMCE developers are there.

I have no device to test the implementation so we have to depend on already known code.

> using your PS1_stats.txt for getting real ids nothing is compromised? [DECKARD config](https://github.com/DKWDRV/DKWDRV/blob/main/docs/files/config_deckard.md)

Haven't had the time to check the output yet, but I will.

> The timestamp strings will increase the filesize a lot unless they can be further shortened down.

I don't think this is a good idea.
In that case might be better off using PSX.EXE CRC32 — title ID table.

**Comment by pcm720** [comment]
@DKWDRV 
I've removed all demo discs, samples and unlicensed titles from that file and ran through each entry manually.
Here's the end result:
https://github.com/pcm720/osdmenu-launcher/blob/main/launcher/include/game_id_table.h

There were a lot of false positives.
I've checked each title by name, all titles present in both this and Deckard config table seem to match.

**Comment by israpps** [comment]
Hi there.



> Does mmceman support all types of mc emus out there? How big is? Stripped down version available? Do we have author permission to integrate it in this project? Does it interfere with normal SIO2MAN, MCMAN drivers? Any special precautions for it?
> 
> I have no device to test the implementation so we have to depend on already known code.
> 
> [@pcm720](https://github.com/pcm720) using your PS1_stats.txt for getting real ids nothing is compromised? [DECKARD config](https://github.com/DKWDRV/DKWDRV/blob/main/docs/files/config_deckard.md)
> 
> The timestamp strings will increase the filesize a lot unless they can be further shortened down.

Project is MIT licensed
https://github.com/ps2-mmce/mmceman/

It hooks into SIO2MAN to transmit into the device more efficiently (early private versions used to hook into SECRMAN to obtain the MCMAN command handler callback)

As for precautions... Not really anything I can think of

You can see the module dependencies on imports.lst

But you'll just need SIO2MAN and FILEXIO

Also, the module already has code for module unloading support, in case DKWDRV is using modern MODLOAD

**Comment by DKWDRV** [comment]
So the final verdict is? Use mmceman for sending game ID?

**Comment by israpps** [comment]
If you don't want to bloat the program with implementation of game history file handling, yes.

It would be as simple as using fileXioDevCtl to issue the appropriate command

**Comment by DKWDRV** [comment]
Is device ping needed?
Game id must be 255 or just strlen of actual id?
Is mcman.irx enough or mmcedrv.irx  is needed aswell?

**Comment by israpps** [comment]
> Is device ping needed?

it is useful to find out wich port has an MMCE device plugged

> Is mcman.irx enough or mmcedrv.irx is needed aswell?

for this use case. mmceman.irx is the only thing you need

mmcedrv.irx is a specialized module intended to be used in the in-game core of OPL for loading ps2 games from MMCE.

> Game id must be 255 or just strlen of actual id?

255 is the max size of the buffer expected by the mmce devices. it can be less, but not more than 255.

**Comment by DKWDRV** [comment]
Does mmceman.irx register "mmcex:" as device? I though it was mmcedrv.irx that did that?

What is the order of loading this module? rom:SIO2MAN, rom0:MCMAN, mmceman.irx?

**Comment by DKWDRV** [comment]
Some users will boot games with no disc scan and that means no ID. Will MCID devices switch/support to PS1 mode?
What if "DWKDRV" is send as ID in all cases when no ID is found?

When sending game id do we have to try and send it to both mmcex0 and mmcex1?

**Comment by DKWDRV** [comment]
Documentation says 255 bytes must be send as game id, if less they must be padded. Does mmceman.irx take care of this padding?

If using rom SIO2MAN and MCMAN, what is the order of loading this module?

**Comment by DKWDRV** [comment]
Added the Volume string to ID table and the driver for sending game id. Do not have a device to test but should work ok.

**Comment by israpps** [comment]
> Does mmceman.irx register "mmcex:" as device? I though it was mmcedrv.irx that did that?
> 
> What is the order of loading this module? rom:SIO2MAN, rom0:MCMAN, mmceman.irx?

MCMAN is not needed

As soon as SIO2MAN is running, mmceman can be loaded

Yes, MMCEMAN is the one registering the `mmce:` device. Mmcedrv will not be of any use on DKWDRV

**Comment by israpps** [comment]
> Some users will boot games with no disc scan and that means no ID. Will MCID devices switch/support to PS1 mode? What if "DWKDRV" is send as ID in all cases when no ID is found?
> 
> When sending game id do we have to try and send it to both mmcex0 and mmcex1?

Sending DKWDRV as gameid will do nothing, because at least Sd2psx will sanity check the format of the gameid

Yes, it would be ideal to send the devctl to both ports. That way you don't have to worry about Wich port has an MMCE device.

**Comment by israpps** [comment]
> Added the Volume string to ID table and the driver for sending game id. Do not have a device to test but should work ok.

I haven't used DKWDRV.
But if you join the PS2 scene server, you'll get access to lots of testers.

**Comment by ZjemCiKolege** [comment]
Game do not detect the card :p

**Comment by DKWDRV** [comment]
> Game do not detect the card :p
> 
> card looks like loaded on sd2psx but in game is not detectable

Don't have a device to test. Report console type PGIF/DECKARD and boot mode CD/USB.

How do you tell if it's loaded or not?

**Comment by ZjemCiKolege** [comment]
> > Game do not detect the card :p
> > card looks like loaded on sd2psx but in game is not detectable
> 
> Don't have a device to test. Report console type PGIF/DECKARD and boot mode CD/USB.
> 
> How do you tell if it's loaded or not?a

deckard usb
works in slot 2 maybe something is in config 

"Memory card path 0" and sd2psx don't work in slot 1 only in 2

**Comment by ZjemCiKolege** [comment]
> [@pcm720](https://github.com/pcm720) using your code to handle both. Is it tested?

Even normal sony ps1 card is not detected in slot 1 ;-D

**Comment by pcm720** [comment]
> @pcm720 using your code to handle both. Is it tested?

@DKWDRV 
Obviously yes, but NHDDL doesn't use the PS1 mode in any way.
I just tested 1.7.6c on my 77000 and it successfully switched both of my SD2PSX into the PS1 mode with VMC mounted.
Cards are flashed with the [latest nightly](https://github.com/sd2psXtd/firmware/commit/f0460f320c57c9ebc7803892792a7c0f9d0e85cd) firmware

It just took a VERY long time to do this for the `mmce1` card when loading both from USB and CD.
I ran a quick test with my own code, and it looks like poll command doesn't work anymore after SD2PSX switches to PS1 mode.
So I suggest just firing the devctl call without checking if it was successful:
```c
char mmceMountpoint[] = "mmceX:";
for (int i = '0'; i < '2'; i++) {
  mmceMountpoint[4] = i;

  // Ping MMCE first to make sure the device is present
  if (fileXioDevctl(mmceMountpoint, 0x1, NULL, 0, NULL, 0) < 0)
    continue;

  fileXioDevctl(mmceMountpoint, 0x8, titleID, (strlen(titleID) + 1), NULL, 0);
}
```
With this code, the switch is pretty much instantaneous.

**Comment by DKWDRV** [comment]
@pcm720 will edit the code to remove the polling command.

With 1.7.6b what happens in 39k? Memory card slot 1 works? @ZjemCiKolege same question for real PS1 mc in slot 1. 

@pcm720 What does Disc License show (I edited some license checks in 1.7.6c)

Anyone with ideas why slot 1 fails?

**Comment by pcm720** [comment]
@DKWDRV 
> With 1.7.6b what happens in 39k? Memory card slot 1 works? @ZjemCiKolege same question for real PS1 mc in slot 1.

Nope, same issue with both SD2PSX in PS1 mode and a regular PS1 memory card:

<details>
  <summary>Gran Turismo 2 Arcade</summary>

  ![Image](https://github.com/user-attachments/assets/039af04e-794d-4e08-b65c-fa873a900974)
</details>

SD2PSX (in PS1 mode) and PS1 memory card in slot 2 work.

> What does Disc License show (I edited some license checks in 1.7.6c)

<details>
  <summary>1.7.6b</summary>

   ![Image](https://github.com/user-attachments/assets/46453b95-9413-4ebb-a259-4bfa51102cad)
</details>

<details>
  <summary>1.7.6c</summary>

  ![Image](https://github.com/user-attachments/assets/ca2c9a66-1746-44d3-9f79-57378f454836)
  ![Image](https://github.com/user-attachments/assets/3ce4b30b-1815-46d4-8250-eacbd10d0c54)
</details>

These issues happen on both unmodified 39k and MechaPwned 77k

**Comment by DKWDRV** [comment]
1.7.6 seems to have an issue with license buffer check. Trying to see what I broke.

What if there is only normal Ps1 card and nothing else? 1.7.6b,c do they work fine?
With SD devices can you you try Launch Original ELF from both version?

Atm no idea what can cause this strange issue of slot 1 not working.

**Comment by pcm720** [comment]
> What if there is only normal Ps1 card and nothing else? 1.7.6b,c do they work fine?
> Atm no idea what can cause this strange issue of slot 1 not working.

I've played around some more and it dawned on me that I've never tried it with the OEM wired controller.
And wouldn't you know it? It works perfectly fine. So was the BlueRetro adapter all along.

Found the relevant issue in BR repo:
https://github.com/darthcloud/BlueRetro/issues/1170

When using the original wired controller, I have _zero_ issues accessing the memory card in slot 1.
Doesn't matter if this card is SD2PSX or a normal PS1 card.
Sorry for wasting your time on this.

So once a new version without the mmceman polling releases, I'm pretty sure we can consider this feature request complete :)

**Comment by ZjemCiKolege** [comment]
> > A co jeśli jest tylko normalna karta Ps1 i nic więcej? 1.7.6b,c czy działają dobrze? 
> > Nie mam pojęcia co może powodować ten dziwny problem z niedziałającym slotem 1.
> 
> Bawiłem się nim jeszcze trochę i dotarło do mnie, że nigdy nie próbowałem go z przewodowym kontrolerem OEM. I czyż nie? Działa idealnie. Podobnie jak adapter BlueRetro przez cały czas.
> 
> Znaleziono odpowiedni problem w repozytorium BR: [darthcloud/BlueRetro#1170](https://github.com/darthcloud/BlueRetro/issues/1170)
> 
> Używając oryginalnego przewodowego kontrolera, nie mam _żadnych_ problemów z dostępem do karty pamięci w slocie 1. Nie ma znaczenia, czy jest to karta SD2PSX czy zwykła karta PS1. Przepraszam za zmarnowanie Twojego czasu.
> 
> Więc gdy tylko ukaże się nowa wersja bez sondowania mmceman, jestem pewien, że będziemy mogli uznać tę prośbę o dodanie funkcji za zakończoną :)

Thanks also, I tested with the blueretro adapter and not the oem pad :P

**Comment by rykotxet** [comment]
Hi! I've done some quick tests with v1.7.6c.

My setup:
- PS2 SCPH-90004 (unmodified)
- SD2PSX with [latest nightly unofficial firmware](https://github.com/sd2psXtd/firmware/releases/tag/nightly) (f0460f3)
- [8bitdo Retro receiver for PS](https://www.8bitdo.com/retro-receiver-for-ps/) paired with an official Dual Shock 4 v2 in the first controller port (if it's relevant)
- SanDisk Ultra 3.0 formatted FAT32/32kb, defragmented and with proper game structure

What I noticed:
- SD2PSX loads the proper PS1 MC for the game launched from USB with DKWDRV (I tested like 10 different games).
- SD2PSX's screen shows the proper PS1 MC mounted within 1 second after pressing X button to boot a USB game.
- When the SD2PSX is plugged in, if I launch a USB game, DKWDRV needs almost 40 seconds to load the game (elapsed time from pressing X to boot a USB game until the PlayStation splash screen finally shows up).
- Same behavior if SD2PSX is plugged in the second memory card slot instead of the first slot.
- When the SD2PSX is NOT plugged in, if I launch a USB game, DKWDRV needs less than 10 seconds to load the game (elapsed time from pressing X to boot a USB game until the PlayStation splash screen finally shows up).
- Saving and loading savegames in SD2PSX's autodetected PS1 MC works fine, as far I tested.

So, there's something that makes game loading last about 30 seconds more when the SD2PSX is plugged, even if SD2PSX loads the PS1 MC in less than a second. That's the only evident issue I noticed.
